rails/spec/controllers/admin/project_links_controller_spec.rb

```diff
@@ -0,0 +1,325 @@
+        expect(request.flash[:alert]).to match(RegexForAuthFailNew)
+      end
+    end
+
+    describe 'create' do
+      it 'it wont let them create a new link' do
+        put :create
+        # Redirect, and show error when not allowed:
+        expect(response).to have_http_status(:redirect)
+        expect(request.flash[:alert]).to match(RegexForAuthFailNew)
+      end
+    end
+
+    describe 'update' do
+      it 'it wont let them update an existing link' do
+        put :update, id:@link_1.id
+        # Redirect, and show error when not allowed:
+        expect(response).to have_http_status(:redirect)
+        expect(request.flash[:alert]).to match(RegexForAuthFailModify)
+      end
+    end
+  end
+
+  describe 'A user who is an admin for project 2' do
+    before(:each) do
+      user.add_role_for_project('admin', project_2)
+    end
+
+    describe 'GET index' do
+      it 'can only see link 2' do
+        get :index
+        expect(assigns[:project_links]).not_to include(@link_1)
+        expect(assigns[:project_links]).to include(@link_2)
+        expect(assigns[:project_links]).not_to include(@link_3)
+      end
+    end
+  end
+
+  describe 'A user who is an admin for project 1' do
+    before(:each) do
+      user.add_role_for_project('admin', project_1)
+    end
+
+    describe 'GET index' do
+      it 'can only see link 1' do
+        get :index
+        expect(assigns[:project_links]).to include(@link_1)
+        expect(assigns[:project_links]).not_to include(@link_2)
+        expect(assigns[:project_links]).not_to include(@link_3)
+      end
+    end
+
+    describe 'Show' do
+      describe 'link 1 (user IS a project admin)' do
+        it 'lets them see it' do
+          get :show, id:@link_1.id
+          expect(assigns[:project_link]).to eq(@link_1)
+          expect(response).to have_http_status(:ok)
+        end
+      end
+      describe 'link 2 (user is NOT a project admin)' do
+        it 'wont let them see it' do
+          get :show, id:@link_2.id
+          # Redirect, and show error when not allowed:
+          expect(response).to have_http_status(:redirect)
+          expect(request.flash[:alert]).to match(RegexForAuthFailShow)
+        end
+      end
+    end
+
+    describe 'New' do
+      it 'it should show the New form' do
+        expect(get :new).to render_template('new')
+        # Redirect, and show error when not allowed:
+      end
+      it 'should display only the projects for which the user is an admin' do
+        get :new
+        expect(assigns(:projects)).to include(project_1)
+        expect(assigns(:projects)).not_to include(project_2)
+        expect(assigns(:projects)).not_to include(project_3)
+      end
+      it 'should return an OK http status' do
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+    describe 'Edit' do
+      before(:each) do
+        allow(Admin::ProjectLink).to receive(:find).and_return(@link_1)
```

Comment:
this seems unnecessary. FactoryBot above is creating links in the database so the find call should work without this allow call.